thumbfast.lua

```diff
@@ -81,6 +81,95 @@ function subprocess(args, async, callback)
     end
 end
 
+local all_processes = {}
+local process_queue = {}
+local active_processes = 0
+local max_processes = 5
+local ytdl_subprocess_cancel = nil
+
+local function spawn_one(args, callback)
+    if active_processes < max_processes then
+        active_processes = active_processes + 1
+        local abort = subprocess(args, true, callback)
+        for i, process in ipairs(process_queue) do
+            if process.args == args then
+                table.remove(process_queue, i)
+                break
+            end
+        end
+        for i, process in ipairs(all_processes) do
+            if process.args == args then
+                all_processes[i] = abort
+                break
+            end
+        end
+    end
+end
+
+local function spawn_queued_process(args, callback)
+    local function wrap_callback(callback)
+        return function(...)
+            callback(...)
+            active_processes = active_processes - 1
+            if #process_queue > 0 then
+                spawn_one(process_queue[1].args, process_queue[1].callback)
+            end
+        end
+    end
+    local wrapped_callback = wrap_callback(callback)
+    local process = {args=args, callback=wrapped_callback}
+    table.insert(process_queue, process)
+    table.insert(all_processes, process)
+    spawn_one(args, wrapped_callback)
+end
+
+local function prioritize_process(atlas_index)
+    local target_process = all_processes[atlas_index]
+    if not target_process or not target_process.args then return end
+
+    for i, process in ipairs(process_queue) do
+        if process.args == target_process.args then
+            if i ~= 1 then
+                table.remove(process_queue, i)
+                table.insert(process_queue, 1, process)
+            end
+            break
+        end
+    end
+end
+
+local function cancel_queued_processes()
+    for i, process in ipairs(all_processes) do
+        if process and not process.args then
+            mp.abort_async_command(process)
+        end
+    end
+    all_processes = {}
+    process_queue = {}
+    if ytdl_subprocess_cancel ~= nil then
+        mp.abort_async_command(ytdl_subprocess_cancel)
+        ytdl_subprocess_cancel = nil
+    end
+end
+
+local function find_closest_index(tbl, target)
+    local lower = target - 1
+    local upper = target + 1
+
+    while lower >= 1 or upper <= #tbl do
+        if tbl[lower] ~= nil then
+            return lower
+        end
+        if tbl[upper] ~= nil then
+            return upper
+        end
+        lower = lower - 1
+        upper = upper + 1
+    end
+
+    return nil
+end
```

Comment:
Shouldn't `tbl[target]` be checked at some point?
```suggestion
local function find_closest_index(tbl, target)
    if tbl[target] ~= nil then
        return target
    end

    local lower = target - 1
    local upper = target + 1

    while lower >= 1 or upper <= #tbl do
        if tbl[lower] ~= nil then
            return lower
        end
        if tbl[upper] ~= nil then
            return upper
        end
        lower = lower - 1
        upper = upper + 1
    end

    return nil
end
```

Reply:
It's checked before calling this function.

Reply:
My bad then.